### PR TITLE
Refactor/callop call oog

### LINKED ISSUE #940: `CALL` & `CALL` OOG refactor
Currently `callop.rs` and `error_oog_call.rs` (and probably more to come) are sharing an important amount of code.
It will be nice to refactor both.

### PR DESCRIPTION
**Pull request by KimiWu123**
fix  #940

### CODE REVIEW

**Comment by ChihChengLiang** [comment]
@DreamWuGit The PR is ready for review. Are you interested in giving this a review?

File: zkevm-circuits/src/evm_circuit/execution/callop.rs
Comment (ChihChengLiang) [inline zkevm-circuits/src/evm_circuit/execution/callop.rs:110]: Are `callee_address_expr` and `code_address` the same?

File: zkevm-circuits/src/evm_circuit/util/common_gadget.rs
Comment (ChihChengLiang) [inline zkevm-circuits/src/evm_circuit/util/common_gadget.rs:505]: If the opcode is not `call` or `callcode`, then we don't pop `value` arg?
```suggestion
```

File: zkevm-circuits/src/evm_circuit/util/common_gadget.rs
Comment (ChihChengLiang) [inline zkevm-circuits/src/evm_circuit/util/common_gadget.rs:502]:
```suggestion
            // `CALL` and `CALLCODE` opcodes have an additional stack pop `value`.
```

File: zkevm-circuits/src/evm_circuit/util/common_gadget.rs
Comment (ChihChengLiang) [inline zkevm-circuits/src/evm_circuit/util/common_gadget.rs:511]: The `CALL_OP` arg is set to false in `error_oog_call.rs` and set to true in `callop.rs`.
We should rename `CALL_OP` to `IS_ERROR` because that differentiates `callop.rs` and  `error_oog_call.rs`.

**Review by ChihChengLiang — COMMENTED** [review]
First pass

File: zkevm-circuits/src/evm_circuit/execution/callop.rs
Comment (KimiWu123) [reply zkevm-circuits/src/evm_circuit/execution/callop.rs:110]: From [the spec](https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/specs/opcode/F1CALL_F2CALLCODE_F4DELEGATECALL_FASTATICCALL.md#evm-behavior), the second arg should be callee address, not code address. I guess in the most of cases, it's a contract. So, the author name it `code_address`.

File: zkevm-circuits/src/evm_circuit/util/common_gadget.rs
Comment (KimiWu123) [reply zkevm-circuits/src/evm_circuit/util/common_gadget.rs:505]: From [the spec](https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/specs/opcode/F1CALL_F2CALLCODE_F4DELEGATECALL_FASTATICCALL.md#evm-behavior), 
```
value - The amount of ether to be transferred (non-existent for both DELEGATECALL and STATICCALL opcodes)
```
only pop `value` while it's `call` and `callcode`. And we could see the [original implementation, L122-L124](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1023/files#diff-e72a6a818cbda6c14cb551573432e59fe61ce109345dd4c088508a60c27979afL122-L124) is the same.

File: zkevm-circuits/src/evm_circuit/util/common_gadget.rs
Comment (KimiWu123) [reply zkevm-circuits/src/evm_circuit/util/common_gadget.rs]: fixed in [c552d5f](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1023/commits/c552d5f6f74350278101b64bd0ae7f04f21226cb)

**Comment by DreamWuGit** [comment]
> @DreamWuGit The PR is ready for review. Are you interested in giving this a review?

will have a look !!

File: zkevm-circuits/src/evm_circuit/util/common_gadget.rs
Comment (ChihChengLiang) [inline zkevm-circuits/src/evm_circuit/util/common_gadget.rs:530]: We can remove this, too. `has_value` is determined by `is_delegatecall`, not `IS_ERROR`.
```suggestion
        let has_value = select::expr(
                is_delegatecall.expr(),
                0.expr(),
                1.expr() - value_is_zero.expr(),
        );
```

File: zkevm-circuits/src/evm_circuit/util/common_gadget.rs
Comment (ChihChengLiang) [inline zkevm-circuits/src/evm_circuit/util/common_gadget.rs:511]: Nitpick: Can we make `stack_push` appear only once?
```suggestion
        cb.stack_push(if IS_ERROR {0.expr()} else {is_success.expr()});
```

File: zkevm-circuits/src/evm_circuit/execution/callop.rs
Comment (ChihChengLiang) [reply zkevm-circuits/src/evm_circuit/execution/callop.rs:110]: Yeah, that makes sense to me.

File: zkevm-circuits/src/evm_circuit/util/common_gadget.rs
Comment (ChihChengLiang) [reply zkevm-circuits/src/evm_circuit/util/common_gadget.rs]: Fixed in [26ca8e7](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1023/commits/26ca8e77ba4b2294225a247d907d88a458de86cb)

**Review by ChihChengLiang — APPROVED** [review]
Nice work. Many duplicated codes were removed!
LGTM after the comments are addressed.

File: zkevm-circuits/src/evm_circuit/util/common_gadget.rs
Comment (DreamWuGit) [inline zkevm-circuits/src/evm_circuit/util/common_gadget.rs:457]: suggestion:  rename `IS_ERROR` to `IS_OOG_ERROR` we have merged other kind of error(insufficient balance error) into CallOp, in the future will probably merge depth overflow error into it. so make the error more clear !

File: zkevm-circuits/src/evm_circuit/execution/error_oog_call.rs
Comment (DreamWuGit) [inline zkevm-circuits/src/evm_circuit/execution/error_oog_call.rs:72]: CallOp already use `callee_exists` to calculate gas cost, here is also old way (check callee_nonce, balance empty to determine existence), would you like to update it in this PR?

File: zkevm-circuits/src/evm_circuit/execution/callop.rs
Comment (DreamWuGit) [inline zkevm-circuits/src/evm_circuit/execution/callop.rs:198]: remove this line .

**Comment by DreamWuGit** [comment]
overall good to me, I will do second review after rebasing and conflicts resolved.

File: zkevm-circuits/src/evm_circuit/execution/error_oog_call.rs
Comment (KimiWu123) [reply zkevm-circuits/src/evm_circuit/execution/error_oog_call.rs:72]: sure, I'll merge `main` first and request for your review again. Thanks!

File: zkevm-circuits/src/evm_circuit/execution/callop.rs
Comment (KimiWu123) [reply zkevm-circuits/src/evm_circuit/execution/callop.rs]: fixed in [7596a17](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1023/commits/7596a17e82b7f5d61b7611aca985419b4c2bb5a9)

File: zkevm-circuits/src/evm_circuit/util/common_gadget.rs
Comment (KimiWu123) [reply zkevm-circuits/src/evm_circuit/util/common_gadget.rs]: fixed in [7596a17](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1023/commits/7596a17e82b7f5d61b7611aca985419b4c2bb5a9)

File: zkevm-circuits/src/evm_circuit/execution/callop.rs
Comment (DreamWuGit) [inline zkevm-circuits/src/evm_circuit/execution/callop.rs:134]: this constraint seems missing in `CommonCallGadget`, it ensure value is zero when not call/call_code.

Comment (KimiWu123) [reply zkevm-circuits/src/evm_circuit/execution/callop.rs:134]: fixed in [ec10a89](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1023/commits/ec10a8953a6fdfa91f169e2fdfa9a08dbb2ac0ab)

File: zkevm-circuits/src/evm_circuit/execution/error_oog_call.rs
Comment (KimiWu123) [reply zkevm-circuits/src/evm_circuit/execution/error_oog_call.rs]: fixed in [e4f6656](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1023/commits/e4f6656be193e0b54b8c2da7b48a01d8ab7c9c4d)

**Comment by KimiWu123** [comment]
@DreamWuGit , have fixed all the comments from your review. It's ready for review now.

@ChihChengLiang, I merged `main` which includes #1020. I'm afraid I missed anything while code merge. Could you please have a quick review again when you have time?

**Review by ChihChengLiang — APPROVED** [review]
LGTM

File: zkevm-circuits/src/evm_circuit/util/common_gadget.rs
Comment (DreamWuGit) [inline zkevm-circuits/src/evm_circuit/util/common_gadget.rs:463]: suggest use old name `phase2_callee_code_hash` so that we easily know clear which phase the RLC do .

**Review by DreamWuGit — APPROVED** [review]
LGTM!

File: zkevm-circuits/src/evm_circuit/util/common_gadget.rs
Comment (KimiWu123) [reply zkevm-circuits/src/evm_circuit/util/common_gadget.rs]: fixed in [8607418](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1023/commits/86074189be1416a61effd63461a9c938da133f42)

**Pull request merged**
